corgi/core/models.py

```diff
@@ -875,12 +874,25 @@ def get_product_details(variant_names: tuple[str], stream_names: list[str]) -> d
         product_streams = ProductNode.get_product_streams(family)
         product_details["productstreams"].update(product_streams)
 
-        product_variants = ProductNode.get_product_variants(family)
-        product_details["productvariants"].update(product_variants)
+        if not variant_names:
+            # We don't know which variants are the right ones
+            # So we just link all of them
+            # TODO: What can we do instead?
+            product_variants = ProductNode.get_product_variants(family)
+            product_details["productvariants"].update(product_variants)
```

Comment:
I think get_product_details needs to be rewritten something like this:

~~~
def get_product_details(variant_names: tuple[str], stream_names: list[str]) -> dict[str, set[str]]:
    """
    Given stream / variant names from the PCR table, look up all related ProductModel subclasses

    In other words, builds relate to variants and streams through the PCR table
    Components relate to builds, and products / versions relate to variants / streams
    We want to link all build components to their ancestor products, versions, and streams
    """

    product_details: dict[str, set[str]] = {
        "products": set(),
        "productversions": set(),
        "productstreams": set(),
        "productvariants": set(),
        "channels": set(),
    }

    if not variant_names and not stream_names:
        return product_details

    query = ProductNode.objects.get_queryset()
    stream_query = Q(level=MODEL_NODE_LEVEL_MAPPING["ProductStream"], productstream__name__in=stream_names)
    variants_query = Q(level=MODEL_NODE_LEVEL_MAPPING["ProductVariant"], productvariant__name__in=variant_names)
    if stream_names and variant_names:
        query = ProductNode.objects.filter(stream_query | variants_query)
    elif stream_names:
        query = ProductNode.objects.filter(stream_query)
    elif variant_names:
        query= ProductNode.objects.filter(variants_query)

    for pnode in query:
        ancestors = pnode.get_ancestors(include_self=True)

        products = ProductNode.get_products(ancestors)
        product_details["products"].update(products)

        product_versions = ProductNode.get_product_versions(ancestors)
        product_details["productversions"].update(product_versions)

        product_streams = ProductNode.get_product_streams(ancestors)
        product_details["productstreams"].update(product_streams)

        product_variants = ProductNode.get_product_variants(ancestors)
        product_details["productvariants"].update(product_variants)

    # For some build, return a mapping of ProductModel type to all related ProductModel UUIDs
    return product_details
~~~

I updated the query to directly lookup ProductNodes with a ProductVariant level and in the product_variant argument set, so I could removed the aggreation of variant_streams in to the stream_names set.

Also inside the loop I use get_ancestors() instead of get_family() so that we don't traverse leaf nodes from those returned in the query.

Reply:
Yes, this addresses point 3. I don't think any code expect a Component has a variant. Where we're hoping to get to is that if a build has a variant, we can then determine it's CPE for an SBOM. And probably other things in the future.